### PR TITLE
Reporting error diagnostics by using `c3c build --test`

### PR DESCRIPTION
**Pull request by pherrymason**
By calling `c3c build --test` on the project, LSP could get some initial diagnostics showing compiler errors as the user modifies its project.

image

- [x] Run diagnostics on project open and document modifications.
- [x] Delay running diagnostics for X seconds after every document modification.
- [x] Include new option to specify path to `c3c`: `c3c-path`
- [x] Include new option to specify with what delay the diagnosis should be executed: `diagnostics-delay`
- [x] Disable diagnostics:
    - [x] if client does not support it.
    - [x] If using a wrong c3c version (one prior to the fix of absolute paths and missing character number)

Reference: https://github.com/c3lang/c3c/issues/1329

### CODE REVIEW

File: server/internal/lsp/server/Diagnostics.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+					URI:         errorInfo.File,
+					Diagnostics: diagnostics,
+				})
+		}
+
+		state.SetCalculateDiagnostics(false)
+	}
+
+	if delay {
+		s.diagnosticDebounced(runDiagnostics)
```
Comment (pherrymason) [inline server/internal/lsp/server/Diagnostics.go:62]: @nikpivkin you seem to have some more experience in Go than me.
I'm unable to correctly debounce the diagnostic calculation: it should not run it until 2 seconds after last document change.

Server (`s` in this line) is storing an instance of `diagnosticDebounced` function so it does not get recreated every document, so in theory, calling again `s.diagnosticDebounced(runDiagnostics)` cancels previous run if not yet executed.

One can easily recreate this bug by spamming keystrokes in a document, and after 2 seconds of the first key press the diagnostics will run.

File: server/internal/lsp/server/server.go [hunk 2]
```diff
@@ -54,14 +66,25 @@ func NewServer(opts ServerOpts) *Server {
...
+	server := &Server{
+		server:  glspServer,
+		options: opts,
+		version: version,
+
+		state:  &state,
+		parser: &parser,
+		search: search,
+
+		diagnosticDebounced: debounce.New(opts.DiagnosticsDelay * time.Millisecond),
```
Comment (pherrymason) [inline server/internal/lsp/server/server.go:86]: Here I create a debounce function and store it so internal timer is conserved between document changes.

**Review by pherrymason — COMMENTED** [review]
@nikpivkin maybe you can help me here with your go knowledge

File: server/internal/lsp/server/TextDocumentDidChange.go [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+import (
+	"github.com/tliron/glsp"
+	protocol "github.com/tliron/glsp/protocol_3_16"
+)
+
+func (s *Server) TextDocumentDidChange(context *glsp.Context, params *protocol.DidChangeTextDocumentParams) error {
+	s.state.UpdateDocument(params.TextDocument.URI, params.ContentChanges, s.parser)
+
+	s.RunDiagnostics(s.state, context.Notify, true)
```
Comment (pherrymason) [inline server/internal/lsp/server/TextDocumentDidChange.go:11]: This is where, after a Document change, I try to trigger a diagnostic run.

**Comment by joshring** [comment]
In order to fix this locally I moved 

`state.SetCalculateDiagnostics(true)` inside of runDiagnostics

**Comment by joshring** [comment]
though that lead to a regression in behaviour
image

**Comment by nikpivkin** [comment]
Hi @pherrymason ! @joshring is right, since `calculatingDiagnostics` is set when diagnostics is run, subsequent calls to diagnostics cannot override the previous one. Does this boolean flag make sense if debouncing is used?

**Comment by pherrymason** [comment]
You are both right, calculatingDiagnostics is causing the bad behaviour. I removed its usage completely as @nikpivkin is right, it might not be needed if I'm already using a debounce.
Thank you!

**Pull request merged**